pkg/server/bundle/client/client.go

```diff
@@ -72,6 +73,14 @@ func NewClient(config ClientConfig) (Client, error) {
 func (c *client) FetchBundle(ctx context.Context) (*bundleutil.Bundle, error) {
 	resp, err := c.client.Get(c.c.EndpointURL)
 	if err != nil {
+		var hostnameError *x509.HostnameError
+		if errors.As(err, &hostnameError) && c.c.SPIFFEAuth == nil {
+			id, e := spiffeid.FromString(hostnameError.Certificate.URIs[0].String())
+			if e != nil {
+				return nil, errs.New("failed to fetch bundle: %v", err)
+			}
+			return nil, errs.New("failed to fetch bundle, the server certificate contains SPIFFE ID %q, should specify https_spiffe instead of https_web: %v", id, err)
+		}
```

Comment:
Line 78 will panic if the certificate does not contain a URI SAN. Also, the branch when `e != nil` is not necessary since we can fall through to the return statement on line 84:
```suggestion
		var hostnameError *x509.HostnameError
		if errors.As(err, &hostnameError) && c.c.SPIFFEAuth == nil && len(hostnameError.Certificate.URIs) > 0 {
			if id, idErr := spiffeid.FromString(hostnameError.Certificate.URIs[0].String()); idErr == nil {
			    return nil, errs.New("failed to authenticate bundle endpoint using web authentication but the server certificate contains SPIFFE ID %q: maybe use https_spiffe instead of https_web: %v", id, err)
			}
		}
```

Reply:
Thanks for your review, @azdagron! I fixed with https://github.com/spiffe/spire/pull/3395/commits/50a169c285de59479dbd9603869ce541a1bb9c40.